app/src/main/java/ch/epfl/qedit/backend/database/FirebaseDBService.java

```diff
@@ -1,49 +1,180 @@
 package ch.epfl.qedit.backend.database;
 
 import androidx.annotation.NonNull;
-import ch.epfl.qedit.util.BundledData;
+import ch.epfl.qedit.model.Question;
+import ch.epfl.qedit.model.Quiz;
 import ch.epfl.qedit.util.Callback;
 import ch.epfl.qedit.util.Response;
 import com.google.android.gms.tasks.OnCompleteListener;
 import com.google.android.gms.tasks.Task;
 import com.google.firebase.firestore.DocumentSnapshot;
 import com.google.firebase.firestore.FirebaseFirestore;
-import java.util.Objects;
+import com.google.firebase.firestore.QueryDocumentSnapshot;
+import com.google.firebase.firestore.QuerySnapshot;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
 
 public class FirebaseDBService implements DatabaseService {
 
-    private final FirebaseFirestore db;
+    private FirebaseFirestore db;
 
     public FirebaseDBService() {
         // Access a Cloud Firestore instance
         db = FirebaseFirestore.getInstance();
     }
 
+    private Question getQuestionFromDoc(QueryDocumentSnapshot doc) throws Exception {
+
+        String title = doc.get("title", String.class);
+        if (title == null) throw new Exception("Title not found in question document");
+        String text = doc.get("text", String.class);
+        if (text == null) throw new Exception("Text not found in question document");
+        String format = doc.get("answer_format", String.class);
+        if (format == null) throw new Exception("Answer format not found in question document");
+
+        return new Question(title, text, format);
+    }
+
     @Override
-    public void getBundle(
-            final String collection,
-            String document,
-            final Callback<Response<BundledData>> responseCallback) {
-        db.collection(collection)
-                .document(document)
+    public void getQuizQuestions(
```

Comment:
I will try to refactor factor out the common lines of code from the three accesses to the database, because there are a lot of things in common. But you have already spent a lot of time trying to simplify it, so it is probably useless...

Reply:
Ok, I saw that we use a `QuerySnapshot` in one method and a `DocumentSnapshot` in the other, which means we cannot really factorize the code.

Reply:
Yes, you have followed the path of my deseperated thought young padawan